### PR TITLE
USB Mass Storage and generic SPI Flash

### PR DESCRIPTION
**Pull request by mmoskal**
@jamesadevine @tballmsft @finneyj  As discussed on call, this PR doesn't include SNORFS, just MSC and generic SPI. 

To summarize, this implements the following:
* various USB interface fixes (both here and `codal-samd21`; see https://github.com/lancaster-university/codal-samd21/pull/1)
* generic SPI flash interface
* implementation of USB Mass Storage Class
* FAT file system implemented virtually (called Ghost FAT, since Virtual FAT is something else)

I want this merged soon, since I want to add WebUSB support, which requires at least the USB fixes from here.

CC @pelikhan @abchatra

### CODE REVIEW

File: source/driver-models/Timer.cpp [hunk 1]
```diff
@@ -225,6 +225,7 @@ int Timer::eventEveryUs(CODAL_TIMESTAMP period, uint16_t id, uint16_t value)
  */
 void Timer::sync(CODAL_TIMESTAMP t)
 {
+    if (t > 3600000000LL) for(;;);
```
Comment (jamesadevine) [inline source/driver-models/Timer.cpp]: Is this required @mmoskal ? Wouldn't it "just wrap around"?

Comment (mmoskal) [reply source/driver-models/Timer.cpp]: This is left-over debugging. This code sometimes does indeed trigger, due to some races somewhere. We get a negative number, which results in a number larger than 63 bits (so much larger than an hour, stated here). This is particularly bad if you use the CPX-specific timer, which only disables timer IRQ not all IRQs. We don't use that timer in PXT though.

Anyway, will remove.

**Comment by jamesadevine** [comment]
Right - so we have a bug in the Timer still? 😟 

Thank you for removing auto from your code :smile:, happy to merge if you are?

**Comment by mmoskal** [comment]
Let me see if I can pile in the WebUSB support code today as well...

**Comment by mmoskal** [comment]
Or actually, let's just merge it and I do WebUSB separately

**Pull request merged**